tests/test_modeling_common.py

```diff
@@ -693,6 +693,99 @@ def test_forward_signature(self):
+                )
+                self.assertTrue(
+                    torch.allclose(batched_row, single_row_object, atol=atol),
+                    msg=(
+                        f"Batched and Single row outputs are not equal in {model_name} for key={key}. Difference="
+                        f"{torch.max(torch.abs(batched_row - single_row_object))}."
+                    ),
+                )
+
+        config, batched_input = self.model_tester.prepare_config_and_inputs_for_common()
+        conv_attr = ["num_channels", "conv_dim", "backbone", "conv_depthwise_kernel_size"]
+        atol = 0.1 if any([hasattr(config, attr) for attr in conv_attr]) else 1e-05
+
+        for model_class in self.all_model_classes:
+            config.output_hidden_states = True
+            config.output_attentions = True
+
+            model_name = model_class.__name__
+            if hasattr(self.model_tester, "prepare_config_and_inputs_for_model_class"):
+                config, batched_input = self.model_tester.prepare_config_and_inputs_for_model_class(model_class)
+            batched_input_prepared = self._prepare_for_class(batched_input, model_class)
+            model = model_class(config).to(torch_device).eval()
+
+            main_input_name = (
+                model.main_input_name[0] if isinstance(model.main_input_name, list) else model.main_input_name
+            )
+            tester_attributes = [i for i in self.model_tester.__dict__.keys() if i[:1] != "_"]
+            model_testers_parents = [attr for attr in tester_attributes if "model_tester" in attr]
+            if model_testers_parents:
+                batch_size = getattr(self.model_tester, model_testers_parents[0]).batch_size
+            elif hasattr(self.model_tester, "batch_size"):
+                batch_size = self.model_tester.batch_size
+            else:
+                batch_size = batched_input_prepared[main_input_name].shape[0]
+
+            single_row_input = {}
+            for key, value in batched_input_prepared.items():
+                if "head_mask" in key:
+                    single_row_input[key] = value
+                elif isinstance(value, torch.Tensor) and value.shape[0] == batch_size:
+                    single_row_input[key] = value[:1]
+                elif isinstance(value, torch.Tensor) and value.shape[0] % batch_size == 0:
+                    # e.g. musicgen has inputs of size (bs*codebooks)
+                    single_batch_shape = value.shape[0] // batch_size
+                    single_row_input[key] = value[:single_batch_shape]
+                else:
+                    single_row_input[key] = value
+
+            with torch.no_grad():
+                model_batched_output = model(**batched_input_prepared)
+                model_row_output = model(**single_row_input)
+
+            if not isinstance(model_batched_output, torch.Tensor):
```

Comment:
Are there cases where the output is a `torch.Tensor`? Even with `config.return_dict = True`?

If so, we should remove the `key` argument in `recursive_check`, to make things easier (and check all models)

Reply:
Yes, FlavaImageCodebook and SpeechT5HifiGan both return a tensor (not tuple of tensors) and do not have options for `return_dict`. I can change it, if it there isn't any special purpose for having tensor outputs

Reply:
Uhmmm perhaps we should convert tensors to dicts before this line, so we check all classes

```py
if not isinstance(model_batched_output, torch.Tensor):
    model_batched_output = {"model_output": model_batched_output}
    model_row_output = {"model_output": model_row_output}
for key in model_batched_output:
    recursive_check(model_batched_output[key], model_row_output[key], model_name, key)
```

Reply:
done